proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,478 @@
+create internal targets. Package traits offer another solution to this without
+requiring additional modules.
+
+### Pluggable dependencies
+
+Some packages want to make it configurable what underlying technology is used.
+The [Swift OpenAPIGenerator](https://github.com/apple/swift-openapi-generator)
+for example is capable of running on top of `URLSession`, `AsyncHTTPClient`,
+`Hummingbird` or `Vapor`. To avoid bringing all of those potential dependencies
+into every adopters binary, the project has created individual repositories for
+each transport. This achieves the goal of making the dependencies optional;
+however, it requires users to look at multiple repositories.
+
+### Configurable behavior
+
+Packages often want to cater to multiple ecosystems such as the iOS or the
+server community. While most of the technologies are shared between ecosystems
+there are often some platform specific behaviors/libraries that one might use.
+For example, on Apple's platforms `OSLog` is the canonical logging systems
+whereas the server ecosystem is mostly using `swift-log`. However, there are
+some users that prefer to use `swift-log` on Apple's platforms which means
+libraries and applications cannot use platform compiler conditionals. Package
+traits would allow to make the logging system configurable independent from the
+platform.
+
+### Replacing environment variables in Package manifests
+
+A lot of packages are using environment variables in their `Package.swift` to
+configure their package. This has various reasons such as optional dependencies
+or setting certain defines for local development. Using environment variables
+inside `Package.swift` is not officially supported and with stricter sandboxing
+rules might break. Package traits provide provide a blessed way to configure a
+package.
+
+## Proposed solution
+
+This proposal introduces a new configuration for packages called _package
+traits_. Package authors can define a set of traits in their `Package.swift`
+that their package offers which provide a way to express conditional compilation
+and optional dependencies. Furthermore, a set of default enabled traits can be
+specified.
+
+```swift
+let package = Package(
+    name: "Example",
+    traits: [
+        "Foo",
+        Trait(
+            name: "Bar",
+            enabledTraits: [ // Other traits that are enabled when this trait is being enabled
+                "Foo"
+            ]
+        )
+    ],
+    defaultTraits: [ // The set of default enabled traits
+        "Foo",
+    ],
+    /// ...
+)
+```
+
+When depending on a package a set of enabled traits can be passed and wether the
+default set of traits are disabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: ["SomeTrait"],
+            disableDefaultTraits: true
```

Comment:
@MaxDesiatov was suggesting something similar. We can certainly make that work. I haven't done it yet since rewiring all of this requires tens of trivial changes :D

Reply:
+1 to what @al45tair proposed. For me personally, singular `.default` reads slightly better.

Reply:
Another +1 from me :)

Reply:
I have updated the proposal to use `.defaults` and disallowed traits to be named `default` or `defaults` (in any casing combination).